src/fem/op_factory.hh

```diff
@@ -0,0 +1,903 @@
+     * the @p p bulk quadrature point.
+     *
+     * @param component_idx Number of the shape function.
+     */
+    inline FeQArray<Scalar> scalar_shape(uint component_idx = 0)
+    {
+        PatchOp<3> *op = this->patch_fe_values_.template get< Op::Bulk::Pt::OpScalarShape<dim, 3> >(dim, component_idx);
+        return FeQArray<Scalar>(op, true);
+    }
+
+    inline FeQArray<Vector> vector_shape(uint component_idx = 0)
+    {
+        auto fe_component = this->fe_comp(fe_, component_idx);
+
+        uint n_dofs = fe_component->n_dofs();
+        uint vector_shape_op_idx = FeBulk::BulkOps::opVectorShape;
+
+        switch (fe_component->fe_type()) {
+            case FEVector:
+            {
+                patch_point_vals_->make_fe_op(vector_shape_op_idx, {3, n_dofs}, bulk_reinit::ptop_vector_shape, n_dofs);
+                break;
+            }
+            case FEVectorContravariant:
+            {
+                ASSERT_PERMANENT(false).error("Shape vector for FEVectorContravariant is not implemented yet!\n"); // temporary assert
+                patch_point_vals_->make_fe_op(vector_shape_op_idx, {3, n_dofs}, bulk_reinit::ptop_vector_contravariant_shape, n_dofs);
+                break;
+            }
+            case FEVectorPiola:
+            {
+                ASSERT_PERMANENT(false).error("Shape vector for FEVectorPiola is not implemented yet!\n"); // temporary assert
+                patch_point_vals_->make_fe_op(vector_shape_op_idx, {3, n_dofs}, bulk_reinit::ptop_vector_piola_shape, n_dofs);
+                break;
+            }
+            default:
+                ASSERT(false).error("Type of FiniteElement of grad_vector_shape accessor must be FEVector, FEVectorPiola or FEVectorContravariant!\n");
+        }
+
+        return FeQArray<Vector>(patch_point_vals_, true, vector_shape_op_idx, fe_component->n_dofs());
+    }
+
+//    inline FeQArray<Tensor> tensor_shape(uint component_idx = 0)
+//    {}
+
+    /**
+     * @brief Return the value of the @p function_no-th gradient shape function at
+     * the @p p bulk quadrature point.
+     *
+     * @param component_idx Number of the shape function.
+     */
+//    inline FeQArray<Vector> grad_scalar_shape(uint component_idx=0)
+//    {
+//        auto fe_component = this->fe_comp(fe_, component_idx);
+//        ASSERT_EQ(fe_component->fe_type(), FEType::FEScalar).error("Type of FiniteElement of grad_scalar_shape accessor must be FEScalar!\n");
+//
+//        uint n_dofs = fe_component->n_dofs();
+//        patch_point_vals_->make_fe_op(FeBulk::BulkOps::opGradScalarShape, {3, n_dofs}, bulk_reinit::ptop_scalar_shape_grads<dim>, n_dofs);
+//
+//        return FeQArray<Vector>(patch_point_vals_, true, FeBulk::BulkOps::opGradScalarShape, n_dofs);
+//    }
+    inline FeQArray<Vector> grad_scalar_shape(uint component_idx=0)
+    {
+        PatchOp<3> *op = this->patch_fe_values_.template get< Op::Bulk::Pt::OpGradScalarShape<dim, 3> >(dim, component_idx);
```

Comment:
Zkusit vyrobit pomocnou factory metodu v předkovi BulkValues, SideValues. Možná některé metody půjdou přesunout do předka celé.

Reply:
Factory sice kód o něco zjednoduší, ale nezbaví nás konstrukce s template parametrem v metodách pro konstrukci ElQ a FeQ accessorů. Factory má template parameter templated class a konstrukce vypadá přibližně následovně:

inline FeQArray<Vector> grad_scalar_shape(uint component_idx=0)
{
    return FeQArray<Vector>( this->template make_op< Op::Bulk::Pt::OpGradScalarShape<dim, 3>  >(component_idx) );
}